Sprint 08

kover opened the pull request:
- Project configuration
- Assets
- Base table and cell layouts

kover committed:
Initial project setup

- Project configuration
- Assets
- Base table and cell layouts

kover committed:
Sprint 08

Adding gradient

Nikitaaeee commented on Imaginarium/Extensions/ImageListViewController+UITableViewDataSource.swift:
```diff
@@ -0,0 +1,28 @@
+
+import UIKit
+
+extension ImagesListViewController: UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 20
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: ImagesListCell.reuseIdentifier, for: indexPath)
+        
+        guard let imageListCell = cell as? ImagesListCell else {
+            return UITableViewCell()
+        }
+        
+        configureCell(for: imageListCell, with: indexPath)
+        return imageListCell
+    }
+
+    
+    
+}
```
Расширения для класса можно держать в самом файле с классом (добавляя все `extension` после самого класса, оставляя комментарии `//MARK: - ***` о том, какой протокол реализует каждый из `extension`). Таким образом, открывая класс, нам будет удобно сразу видеть дополнительный функционал, который у него есть. 

Можно прочитать про это в гайдлайнах по кодстайлу Swift (раздел Protocol Conformance):
https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization

Обычно в папке Extensions хранят расширения для системных классов (например UIColor+Extension, или Date+Extension)

kover replied:
Понял, спасибо. Исправляю

Nikitaaeee commented on Imaginarium/Extensions/ImageListViewController+CellConfiguration.swift:
```diff
@@ -0,0 +1,24 @@
+
+import UIKit
+
+extension ImagesListViewController {
+    func configureCell(for cell: ImagesListCell, with indexPath: IndexPath) {
+        guard let image = UIImage(named: photosName[indexPath.row]) else {
+            return
+        }
+        
+        cell.setCellImage(image)
+        
+        cell.setPostDate(dateFormatter.string(from: Date()))
+
+        cell.isFavorite = indexPath.row % 2 == 0
+        
+        cell.addGradient()
+    }
+}
```
Для этого метода стоит добавить `private` (после того, как перенесешь этот и остальные extension's в файл с классом `ImagesListViewController`, как я описал в другом комментарии)
Private необходимо указывать  для того, чтобы спрятать детали реализации конкретной части функциональности,
когда они используются только внутри области объявления.

Здесь и далее в проектах, не забывай проставлять private всем свойствам и функциям, которые не планируешь использовать вне класса.

Если у нас будет много приватных методов, то мы можем сделать приватный extension, тогда все методы внутри этого `private extension` автоматически станут приватными и нам не нужно будет прописывать private перед ними.

kover replied:
Согласен, поправил.

Nikitaaeee commented on Imaginarium/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,30 @@
+
+import UIKit
+
+class ImagesListViewController: UIViewController {
+
+    let photosName: [String] = Array(0..<20).map{ "\($0)" }
```
Для улучшения производительности можно добавлять модификатор `final` к классам, от которых мы не будем наследоваться. (Как ты сделал в `ImagesListCell`, например)
В таком случае комплятор будет быстрее находить адрес ячейки памяти при вызове функции.

Если интересно, тут можно подробнее почитать про то, что такое диспетчеризация и как она работает (довольно частая тема на собеседованиях):
https://habr.com/ru/companies/simbirsoft/articles/673636/

kover replied:
Спасибо за ссылку, поправил.

Nikitaaeee commented on Imaginarium/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,30 @@
+
+    let photosName: [String] = Array(0..<20).map{ "\($0)" }
+    
+    lazy var dateFormatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateStyle = .long
+        formatter.timeStyle = .none
+        return formatter
+    }()
+    
```
Для форматтера тоже можно добавить private, если он не будет использоваться извне.

kover replied:
Сделал.

Nikitaaeee commented on Imaginarium/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,63 @@
+    @IBOutlet private var postDate: UILabel!
+    @IBOutlet private var gradientView: UIView!
+    
+    @IBAction func toggleLike() {
+        isFavorite = !isFavorite
+    }
+    
```
Добавить интерактивность кнопке лайка - крутая идея!

Nikitaaeee commented on Imaginarium/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,63 @@
+        postDate.text = date
+    }
+    
+    func addGradient() {
+        gradient.frame = gradientView.bounds
+        gradientView.layer.addSublayer(gradient)
+    }
```
Градиент добавлен - доп. задание выполнено 🔥!

Nikitaaeee commented on Imaginarium/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,63 @@
+    
+    static let reuseIdentifier = "ImagesListCell"
+    
+    private var _isFavorite = true
+    
```
Называть булевые переменные через is*** - хорошая практика, но в кодстайле Swift обычно не используются префиксы "_", поэтому можно его убрать.

kover replied:
Поправил

Nikitaaeee commented on Imaginarium/ImagesList/ImagesListCell.swift:
```diff
@@ -0,0 +1,63 @@
+        let gradient = CAGradientLayer()
+        gradient.type = .axial
+        gradient.colors = [
+            UIColor(named: "Gradient Start")!.cgColor,
+            UIColor(named: "Gradient Stop")!.cgColor
```
Использование force unwrapping может привести к ошибкам выполнения, если объект окажется равным nil. 
Можно отдельно задать два свойства (например, startColor и endColor) и заменить принудительное извлечение на безопасное извлечение.

kover replied:
Поправил

kover committed:
Sprint 08 Review changes

- Add table view constraints
- Reorganize extensions
- Update access modifiers
- Update toolbar style

kover committed:
Sprint 08 Review update

- Change table view constraints relation to Superview
